src/outputs/outputs.cpp

```diff
@@ -439,8 +439,17 @@ void Outputs::MakeOutputs(Mesh *pm, ParameterInput *pin, SimTime *tm,
            ptype->output_params.analysis_flag)))) {
       if (first && ptype->output_params.file_type != "hst") {
         pm->ApplyUserWorkBeforeOutput(pm, pin, *tm);
+        for (const auto &pkg : pm->packages.AllPackages()) {
+          pkg.second->UserWorkBeforeOutput(pm, pin, *tm);
+        }
         first = false;
       }
+      if (ptype->output_params.file_type == "rst") {
+        pm->ApplyUserWorkBeforeRestartOutput(pm, pin, *tm, &(ptype->output_params));
+        for (const auto &pkg : pm->packages.AllPackages()) {
+          pkg.second->UserWorkBeforeRestartOutput(pm, pin, *tm, &(ptype->output_params));
+        }
```

Comment:
General question: Do we want to/can we make any guarantees on the order of execution here (on other words, should we expect competing work being done here in downstream codes)?
Anyway, might be worth documenting that what we can guarantee and what not in the dev docs above.

Reply:
I state in the documentation now that the per-mesh callbacks precede the per-package callbacks, but both are always called if they are provided.